src/DataStructures/Mesh.hpp

```diff
@@ -96,17 +97,54 @@ class Mesh : public Index<Dim> {
    */
   Mesh(std::array<size_t, Dim> extents, std::array<Spectral::Basis, Dim> bases,
        std::array<Spectral::Quadrature, Dim> quadratures) noexcept
-      : Index<Dim>(std::move(extents)),
+      : extents_(std::move(extents)),
         bases_(std::move(bases)),
         quadratures_(std::move(quadratures)) {}
 
+  /*!
+   * \brief The number of grid points in each dimension of the grid.
+   */
+  const Index<Dim>& extents() const noexcept { return extents_; }
+
+  /*!
+   * \brief The number of grid points in dimension `d` of the grid.
+   */
+  size_t extents(const size_t d) const noexcept { return extents_[d]; }
+
+  /*!
+   * \brief The total number of grid points in all dimensions.
+   *
+   * \details `DataVector`s that represent field values on the grid have this
+   * many entries.
+   *
+   * For `Dim = 0` the number of grid points is zero.
+   */
+  template <int N = Dim, Requires<(N > 0)> = nullptr>
+  constexpr size_t number_of_grid_points() const noexcept {
+    return extents_.product();
+  }
+  /// \cond
+  // Specialization for `Dim = 0` to return `0`.
+  template <int N = Dim, Requires<(N == 0)> = nullptr>
+  constexpr size_t number_of_grid_points() const noexcept {
+    return 0;
```

Comment:
Why is this special N=0 version here?  The correct value is 1, which would be returned by the general version.

Reply:
Well I figured a zero-dimensional mesh should actually return zero `number_of_grid_points`, this matches the expectation I think. On the other hand, for `Index::product` the meaning in zero dimensions is not quite as well-defined, and could be defined as one. What do you think?

Reply:
a zero-dimensional mesh is the "face" of a 1-d element and has 1 grid point

Reply:
Makes sense :) Fixed.